File: frontend/src/api/jobs.ts
import { buildApiUrl } from './base'
import type { JobCreateRequest, JobLogsResponse, JobResponse } from './generated/contracts'

export type { JobCreateRequest, JobLogsResponse, JobResponse }

export async function createJob(request: JobCreateRequest): Promise<JobResponse> {
  const response = await fetch(buildApiUrl('/jobs'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  })

  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to create job: ${response.status} ${text}`)
  }
  return response.json()
}

export async function getJob(jobId: string): Promise<JobResponse> {
  const response = await fetch(buildApiUrl(`/jobs/${jobId}`))
  if (!response.ok) {
    throw new Error(`Failed to fetch job: ${response.status}`)
  }
  return response.json()
}

export async function listJobs(limit = 30): Promise<JobResponse[]> {
  const response = await fetch(buildApiUrl(`/jobs?limit=${limit}`))
  if (!response.ok) {
    throw new Error(`Failed to list jobs: ${response.status}`)
  }
  return response.json()
}

export async function getJobLogs(jobId: string, tail = 200): Promise<JobLogsResponse> {
  const response = await fetch(buildApiUrl(`/jobs/${jobId}/logs?tail=${tail}`))
  if (!response.ok) {
    throw new Error(`Failed to fetch logs: ${response.status}`)
  }
  return response.json()
}

export async function cancelJob(jobId: string): Promise<JobResponse> {
  const response = await fetch(buildApiUrl(`/jobs/${jobId}/cancel`), {
    method: 'POST',
  })
  if (!response.ok) {
    throw new Error(`Failed to cancel job: ${response.status}`)
  }
  return response.json()
}

File: backend/api/jobs.py
from typing import List

from fastapi import APIRouter, HTTPException, Query

from services.job_runner import job_runner
from schemas.jobs import JobCreateRequest, JobLogsResponse, JobResponse


router = APIRouter(prefix="/api/jobs", tags=["jobs"])


@router.post("", response_model=JobResponse)
def create_job(request: JobCreateRequest):
    payload = request.model_dump()
    try:
        job = job_runner.create_job(payload)
        return JobResponse(**job)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.get("", response_model=List[JobResponse])
def list_jobs(limit: int = Query(default=50, ge=1, le=200)):
    jobs = job_runner.list_jobs(limit=limit)
    return [JobResponse(**item) for item in jobs]


@router.get("/{job_id}", response_model=JobResponse)
def get_job(job_id: str):
    try:
        return JobResponse(**job_runner.get_job(job_id))
    except KeyError:
        raise HTTPException(status_code=404, detail="Job not found")


@router.get("/{job_id}/logs", response_model=JobLogsResponse)
def get_job_logs(job_id: str, tail: int = Query(default=200, ge=10, le=2000)):
    try:
        return JobLogsResponse(**job_runner.get_job_logs(job_id=job_id, tail=tail))
    except KeyError:
        raise HTTPException(status_code=404, detail="Job not found")


@router.post("/{job_id}/cancel", response_model=JobResponse)
def cancel_job(job_id: str):
    try:
        return JobResponse(**job_runner.cancel_job(job_id))
    except KeyError:
        raise HTTPException(status_code=404, detail="Job not found")
